backend/pkg/notification/types.go

```diff
@@ -0,0 +1,864 @@
+	}
+	return ""
+}
+
+func formatDashboardAndGroupLink(format types.NotificationFormat, n types.Notification) string {
+	dashboardAndGroupInfo := ""
+	if n.GetDashboardId() != nil {
+		switch format {
+		case types.NotifciationFormatHtml:
+			dashboardAndGroupInfo = fmt.Sprintf(` of Group <b>%[2]v</b> in Dashboard <a href="https://%[1]v/dashboard/%[4]v">%[3]v</a>`, utils.Config.Frontend.SiteDomain, n.GetDashboardGroupName(), n.GetDashboardName(), *n.GetDashboardId())
+		case types.NotifciationFormatText:
+			dashboardAndGroupInfo = fmt.Sprintf(` of Group %[1]v in Dashboard %[2]v`, n.GetDashboardGroupName(), n.GetDashboardName())
+		case types.NotifciationFormatMarkdown:
+			dashboardAndGroupInfo = fmt.Sprintf(` of Group **%[1]v** in Dashboard [%[2]v](https://%[3]v/dashboard/%[4]v)`, n.GetDashboardGroupName(), n.GetDashboardName(), utils.Config.Frontend.SiteDomain, *n.GetDashboardId())
+		}
+	}
+	return dashboardAndGroupInfo
+}
+
+type ValidatorProposalNotification struct {
+	types.NotificationBaseImpl
+
+	ValidatorIndex uint64
+	Slot           uint64
+	Block          uint64
+	Status         uint64 // * Can be 0 = scheduled, 1 executed, 2 missed */
+	Reward         float64
+}
+
+func (n *ValidatorProposalNotification) GetEntitiyId() string {
+	return fmt.Sprintf("%d", n.ValidatorIndex)
+}
+
+func (n *ValidatorProposalNotification) GetInfo(format types.NotificationFormat) string {
+	vali := formatValidatorLink(format, n.ValidatorIndex)
+	slot := formatSlotLink(format, n.Slot)
+	dashboardAndGroupInfo := formatDashboardAndGroupLink(format, n)
+
+	switch n.Status {
+	case 0:
+		return fmt.Sprintf(`New scheduled block proposal at slot %s for Validator %s%s.`, slot, vali, dashboardAndGroupInfo)
+	case 1:
+		return fmt.Sprintf(`Validator %s%s proposed block at slot %s with %v %v execution reward.`, vali, dashboardAndGroupInfo, slot, n.Reward, utils.Config.Frontend.ElCurrency)
+	case 2:
+		return fmt.Sprintf(`Validator %s%s missed a block proposal at slot %s.`, vali, dashboardAndGroupInfo, slot)
+	case 3:
+		return fmt.Sprintf(`Validator %s%s had an orphaned block proposal at slot %s.`, vali, dashboardAndGroupInfo, slot)
+	default:
+		return "-"
+	}
+}
+
+func (n *ValidatorProposalNotification) GetLegacyInfo() string {
+	var generalPart, suffix string
+	vali := strconv.FormatUint(n.ValidatorIndex, 10)
+	slot := strconv.FormatUint(n.Slot, 10)
+
+	switch n.Status {
+	case 0:
+		generalPart = fmt.Sprintf(`New scheduled block proposal at slot %s for Validator %s.`, slot, vali)
+	case 1:
+		generalPart = fmt.Sprintf(`Validator %s proposed block at slot %s with %v %v execution reward.`, vali, slot, n.Reward, utils.Config.Frontend.ElCurrency)
+	case 2:
+		generalPart = fmt.Sprintf(`Validator %s missed a block proposal at slot %s.`, vali, slot)
+	case 3:
+		generalPart = fmt.Sprintf(`Validator %s had an orphaned block proposal at slot %s.`, vali, slot)
+	}
+	return generalPart + suffix
+}
+
+func (n *ValidatorProposalNotification) GetTitle() string {
```

Comment:
Could you elaborate what you mean here? I don't quite understand your feedback.

Reply:
Sure, I'd suggest doing this instead of using the same function body twice where applicable:
```suggestion
func (n *ValidatorProposalNotification) GetTitle() string {
	return n.GetLegacyTitle()
}
```
It applies to some other functions like `func (n *ValidatorIsOfflineNotification) GetTitle() string`, too.

IMO advantages are it improves readability, reduces file/binary size and clarifies intent (i.e. "new implementation doesn't deviate from legacy implementation yet")